### PR TITLE
Implement Mozlog and remove CEF logging

### PR DESCRIPTION
**Pull request by bhearsum**
This branch implements mozlog for Balrog and fully replaces CEF logging with it. The JsonLogFormatter is taken from https://raw.githubusercontent.com/mozilla/testpilot/master/testpilot/base/logging.py. I'm told it's going to be made available independent at some point, so we'll probably switch to that when it is.

Logging configuration has changed a little bit here as well, with Mozlog formatting being the default, but overridable with an environment variable (which we use for the Docker images).

One thing we're losing here is granularity of errors (CEF_* all become a WARNING here) - should I be doing something different here, @jvehent? My assumption is that we're probably going to be running at log level WARNING in production, so I perhaps I should be switching some of these to ERROR or CRITICAL messages?

There's a couple of new TODOs here. One of them is about e-mail notification for rules/permission changes. I'll be doing that in bug 1251338. The other is a note about the requestid that BalrogLogger generates. That's a new bug I found that I don't intend to fix here.

### CODE REVIEW

File: auslib/log.py [hunk 1]
```diff
@@ -26,42 +26,118 @@ def makeRecord(self, name, level, fn, lno, msg, args, exc_info, func=None, extra
...
+
+    This log formatter outputs JSON format messages that are compatible with
+    Mozilla's standard heka-based log aggregation infrastructure.
+
+    See also:
+    https://mana.mozilla.org/wiki/display/CLOUDSERVICES/Logging+Standard
+    https://mana.mozilla.org/wiki/pages/viewpage.action?pageId=42895640
+
+    Adapted from:
+    https://github.com/mozilla-services/mozservices/blob/master/mozsvc/util.py#L106
```
Comment (nthomas-mozilla) [inline auslib/log.py]: Lets change master to a specific revision, or otherwise note what revision you used.

**Comment by nthomas-mozilla** [comment]
Seems fine to me from a code inspection point of view. I'm curious what the actual logging looks like, and to hear the answer to the levels question. Fair to say we might have missed some runtime errors because test coverage is minimal ?

**Comment by bhearsum** [comment]
Here's some output after commenting out LOG_FORMAT=plain from docker-compose.yml:
```
balrogadmin_1 | {"EnvVersion": "2.0", "Type": "SingleReleaseView", "Severity": 7, "Fields": {"message": "processing PUT request to /releases/B2G-dogfood-fota-20151130130150", "requestid": 139850289015120}, "Timestamp": 1458309773690236928, "Logger": "Balrog", "Hostname": "2abf3b0d38a9", "Pid": 11}
balrogadmin_1 | {"EnvVersion": "2.0", "Type": "Releases", "Severity": 7, "Fields": {"message": "Looking for releases with:", "requestid": 139850289015120}, "Timestamp": 1458309773706931968, "Logger": "Balrog", "Hostname": "2abf3b0d38a9", "Pid": 11}
balrogadmin_1 | {"EnvVersion": "2.0", "Type": "Releases", "Severity": 7, "Fields": {"message": "name: B2G-dogfood-fota-20151130130150", "requestid": 139850289015120}, "Timestamp": 1458309773707062016, "Logger": "Balrog", "Hostname": "2abf3b0d38a9", "Pid": 11}
balrogadmin_1 | {"EnvVersion": "2.0", "Type": "Releases", "Severity": 7, "Fields": {"message": "product: None", "requestid": 139850289015120}, "Timestamp": 1458309773707262976, "Logger": "Balrog", "Hostname": "2abf3b0d38a9", "Pid": 11}
balrogadmin_1 | {"EnvVersion": "2.0", "Type": "AUSTransaction", "Severity": 7, "Fields": {"message": "Attempting to execute SELECT releases.`read_only` \nFROM releases \nWHERE releases.name = %s", "requestid": 139850289015120}, "Timestamp": 1458309773713052928, "Logger": "Balrog", "Hostname": "2abf3b0d38a9", "Pid": 11}
balrogadmin_1 | {"EnvVersion": "2.0", "Type": "auslib.blobs.apprelease.ReleaseBlobV4", "Severity": 7, "Fields": {"requestid": 139850289015120}, "Timestamp": 1458309773713998080, "Logger": "Balrog", "Hostname": "2abf3b0d38a9", "Pid": 11}
balrogadmin_1 | {"EnvVersion": "2.0", "Type": "SingleReleaseView", "Severity": 4, "Fields": {"message": "Bad input: Couldn't update release: name in database (B2G-dogfood-fota-20151130130150) does not match name in blob (B2G-mozilla-b2g37_v2_2-nightly-latest)", "requestid": 139850289015120}, "Timestamp": 1458309773754620160, "Logger": "Balrog", "Hostname": "2abf3b0d38a9", "Pid": 11}
balrogadmin_1 | {"EnvVersion": "2.0", "Type": "AUSTransaction", "Severity": 7, "Fields": {"message": "exc is:", "traceback": "Uncaught exception:\nNone\nNone\n", "requestid": 139850289015120, "error": "None"}, "Timestamp": 1458309773755002112, "Logger": "Balrog", "Hostname": "2abf3b0d38a9", "Pid": 11}
```

I'm not exactly sure what you're getting at re: missing runtime errors. Which test coverage are you referring to being minimal?

File: auslib/db.py [hunk 2]
```diff
@@ -1275,20 +1274,22 @@ def hasUrlPermission(self, username, url, method, urlOptions={}, transaction=Non
...
 
     def onDelete(table, who, where):
         if who not in systemAccounts:
-            cef_event('Human modification', CEF_ALERT, user=who, where=where, table=table.name, type='delete')
+            pass
 
     def onUpdate(table, who, where, what):
         if who not in systemAccounts:
-            cef_event('Human modification', CEF_ALERT, user=who, what=what, where=where, table=table.name, type='update')
+            pass
```
Comment (jvehent) [inline auslib/db.py]: Why not log those as `info` in the meantime?

File: auslib/log.py [hunk 3]
```diff
@@ -26,42 +26,118 @@ def makeRecord(self, name, level, fn, lno, msg, args, exc_info, func=None, extra
                 # Request object, whose id is actually what we want.
                 # Without this we end up with the id of the proxy object, which
                 # is static for the life of the application.
+                # TODO: this doesn't seem to be 100% unique.
+                # Sometimes requests that happen around the same time will
+                # end up with the same value here. Possibly only when the
+                # query strings are the same, or srcip?
                 requestid = id(request._get_current_object())
```
Comment (jvehent) [inline auslib/log.py]: add a random?
```python
>>> iv = os.urandom(8)
>>> print("".join("{:02x}".format(c) for c in iv))
a6ddfb5d76f21422
```

**Comment by jvehent** [comment]
> One thing we're losing here is granularity of errors (CEF_* all become a WARNING here) - should I be doing something different here, @jvehent? My assumption is that we're probably going to be running at log level WARNING in production, so I perhaps I should be switching some of these to ERROR or CRITICAL messages?

@mostlygeek would know better, but my understanding is we log *everything* coming out of the application. If you set the log level to DEBUG, then we'll log at that level. We don't post-filter on log level in the logging pipeline. 
I generally recommend logging at INFO level, and since you have everything set to WARNING, we will be fine there. Log volume isn't a concern.

Changing some of those levels from WARNING to ERROR or CRITICAL is up to you. I think a violation that gets blocked is fine as a WARNING. An application that should have succeeded but failed is an ERROR, and something massively wrong that's causing the app to crash is a CRITICAL (like a DB connection failure, for example). Feel free to adapt those to your liking.

File: auslib/db.py (hunk 2 above)
Comment (bhearsum) [inline auslib/db.py]: Could do, I've been treating email notification as a must have before we transition, so I didn't think it mattered. It's probably worthwhile logging them in additional to e-mail though, so I'll do this.

File: auslib/log.py (hunk 3 above)
Comment (bhearsum) [inline auslib/log.py]: I might be missing something, but I'm not sure how that helps? If I generate a new random number for each call to make_record, each log entry (even from within the same request) will end up with a different id. I don't want to block on this in any case...I just didn't want to forget that this needs to get fixed at some point.

**Comment by bhearsum** [comment]
I'm a bit hesistant to log at DEBUG by default - that enables some _extremely_ spamming stuff in auslib/db.py (hundreds of lines per request).

In any case, it sounds like as long as all these former cef events get logged in some way, we're good. I'll have another look over them to see if some should be switched to ERROR.

**Comment by jvehent** [comment]
> I'm a bit hesistant to log at DEBUG by default - that enables some extremely spamming stuff in auslib/db.py (hundreds of lines per request).

Sorry, that was unclear, I'm absolutely **not** suggesting you log at DEBUG by default, that would be madness :sob: Just log at WARNING or INFO and we'll be fine.

r+ for me

**Comment by mostlygeek** [comment]
For production we should/will have the log level set to INFO.

File: auslib/db.py (hunk 2 above)
Comment (bhearsum) [inline auslib/db.py]: Actually, I just noticed that my email notification patch does this, so I'll leave this stubbed for now. That'll be up for review as soon as this one merges.

**Comment by bhearsum** [comment]
This most recent commit changes to the default log level to match what it'll be in prod. While I was poking at this some more I also noticed that the COPY commands in the Dockerfiles are wrong - they copied the contents of directories such as "auslib" without including the dirname. This was never noticed in dev because of the bind mounts, it would've busted production though.

**Comment by bhearsum** [comment]
Recent landings (especially my Dockerflow patch) have caused some really ugly history. I'm going to force-push a rebased version of this branch to avoid that here.

**Comment by bhearsum** [comment]
I followed up with Nick on IRC about his "test coverage" comment, and I'll be adding a few tests around the JsonLogFormatter to address that. I'm going to merge this in the meantime because it contains Dockerfile fixes that are preventing @mostlygeek from using the current Docker image. I'll send a new PR with the tests when they're ready.

**Pull request merged**